### ISSUE TITLE
Stages

### ISSUE THREAD

**Issue by conneroisu** [conneroisu/sfhw-proj2#32]
- [x] #30
- [x] #31
- [x] #28
- [x] #29

![image](https://github.com/user-attachments/assets/8e8d5e84-ca22-462e-8b85-ea1c00c43e8f)
Diagram Representation:
![image](https://github.com/user-attachments/assets/02470527-897f-4a9c-a4e6-f5846a3722b1)

![image](https://github.com/user-attachments/assets/f7c4c846-9594-4dd0-a29a-425b239bbf25)

Taken from zybooks:

Figure 4.7.2: The pipelined version of the single-cycle datapath (COD Figure 4.35).

The pipeline registers, in color, separate each pipeline stage. They are labeled by the stages that they separate; for example, the first is labeled IF/ID because it separates the instruction fetch and instruction decode stages. The registers must be wide enough to store all the data corresponding to the lines that go through them. For example, the IF/ID register must be 64 bits wide, because it must hold both the 32-bit instruction fetched from memory and the incremented 32-bit PC address. We will expand these registers over the course of this chapter, but for now the other three pipeline registers contain 128, 97, and 64 bits, respectively. (NOTE I THINK THIS IS IGNORING THE COMMON SIGNALS!)
![image](https://github.com/user-attachments/assets/aa948f68-c3de-48b6-9319-33f6a1eccd92)
link: https://learn.zybooks.com/zybook/IASTATECPRE3810GulmezogluFall2024/chapter/4/section/7?content_resource_id=61945231

<sub><a href="https://huly.app/guest/cpre381?token=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJsaW5rSWQiOiI2NzJjZGU0NGU3Y2M1YTc4NTQyOTU1ZmIiLCJndWVzdCI6InRydWUiLCJlbWFpbCI6IiNndWVzdEBoYy5lbmdpbmVlcmluZyIsIndvcmtzcGFjZSI6InctZ2l0aHViY29ubmVyLWNwcmUzODEtNjcyNTBhMGUtNWMwM2NlOGY1Zi02YmU0NzMifQ.DDR6P8JvCUUMBjCNMoXdmfnbWz65VKx_Lv6W10jlo6I">Huly&reg;: <b>PROJE-32</b></a></sub>

**Comment by conneroisu** [comment]
One benefit of designing our processor this way is the increased ability to optimize efficiently, by breaking up our logic into the stages, when we run the timings, we can identify possible bottlenecks from the given path/critical path.

**Comment by conneroisu** [comment]
Another good image showing the values of control signals passed through stages:
![image](https://github.com/user-attachments/assets/10b3d4a8-1cc8-4f86-8a80-ed5447c9c8d9)

**Comment by conneroisu** [comment]
I made this as a document in the stages folder also but here is how I went about creating my stage without component instatiations (like registerfile, alu, etc):
This document is a guide for the design of the stages in the hardware pipelined MIPS processor.

# Stage Guide

1. Use the common signals for each stage (i.e. `i_CLK`, `i_RST`, `i_WE`, `i_PC`, `o_PC`, `i_PCplus4`, and `o_PCplus4`)

2. Add inputs for the stage specific signals (changes per stage)
See: https://private-user-images.githubusercontent.com/88785126/384028866-8e8d5e84-ca22-462e-8b85-ea1c00c43e8f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzEzMzIyOTUsIm5iZiI6MTczMTMzMTk5NSwicGF0aCI6Ii84ODc4NTEyNi8zODQwMjg4NjYtOGU4ZDVlODQtY2EyMi00NjJlLThiODUtZWExYzAwYzQzZThmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTExVDEzMzMxNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTMzNzM4ZjA4NDAxYjVhM2ZhMzQyNzIxNTJjYTBlNTc3ZjRiY2NlZWUwZTFhOWZkMGNhNzliMDVkMDM5MTgyMDUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.XG3ua9fePqVvlV9ENBneOm_dfjEvY2qnAtWhg7wU6xU

3. Add passed through signals for the stage specific signals (changes per stage)
In the table, all passed through signals are stages below the chosen to be implemented stage.
See: https://private-user-images.githubusercontent.com/88785126/384028866-8e8d5e84-ca22-462e-8b85-ea1c00c43e8f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzEzMzIyOTUsIm5iZiI6MTczMTMzMTk5NSwicGF0aCI6Ii84ODc4NTEyNi8zODQwMjg4NjYtOGU4ZDVlODQtY2EyMi00NjJlLThiODUtZWExYzAwYzQzZThmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTExVDEzMzMxNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTMzNzM4ZjA4NDAxYjVhM2ZhMzQyNzIxNTJjYTBlNTc3ZjRiY2NlZWUwZTFhOWZkMGNhNzliMDVkMDM5MTgyMDUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.XG3ua9fePqVvlV9ENBneOm_dfjEvY2qnAtWhg7wU6xU

4. Add inputs from components of the previous stage (changes per stage)

Basically, add the inputs to the stage from the outputs of components from diatgrams of the previous stage.

5. Add outputs to components of the next stage (changes per stage)

6. Add "caching" of stage specific signals (changes per stage)

**Comment by conneroisu** [comment]
I have since updated the stage creation guide and will be pushing it to the repo soon

**Comment by conneroisu** [comment]
Another cool image showing how we plan to comparmentatize all the inner components into the stages:
![image](https://github.com/user-attachments/assets/2be9474a-3a31-49d7-af46-e76c83ab2437)

**Comment by conneroisu** [comment]
![image](https://github.com/user-attachments/assets/916b24ff-bba0-4cf2-8429-ba9ae7f015fd)

**Comment by conneroisu** [comment]
![image](https://github.com/user-attachments/assets/2a09c7cf-c4a6-41e1-91b1-dead0df17b0e)

**Comment by conneroisu** [comment]
![image](https://github.com/user-attachments/assets/35cc4c72-9f1b-46e7-819f-c96601bb7c5d)

Figure 4.7.11: The corrected pipelined datapath with the control signals identified. (COD Figure 4.46).

**Comment by conneroisu** [comment]
![image](https://github.com/user-attachments/assets/a5c09a14-c696-4c1f-b114-093d8452a110)

**Comment by conneroisu** [comment]
https://app.eraser.io/workspace/ONdQZqJ8JixF0Aaz8n6M?origin=share

**Comment by conneroisu** [comment]
![image](https://github.com/user-attachments/assets/ebe47b60-187d-4151-b822-da44cd6b43e2)

**Comment by conneroisu** [comment]
[Proj2_report.pdf](https://github.com/user-attachments/files/17970755/Proj2_report.pdf)
